tdrs-backend/tdpservice/parsers/validators.py

```diff
@@ -172,121 +177,133 @@ def sumIsLargerFunc(value):
             if sum > val
             else (
                 False,
-                f"The sum of {fields} is not larger than {val}.",
+                f"{record_type}: The sum of {fields} is not larger than {val}.",
                 [field for field in fields],
             )
         )
 
-    return lambda value: sumIsLargerFunc(value)
+    return sumIsLargerFunc
 
 
 # generic validators
 
 
-def matches(option):
+def matches(option, error_func=None):
     """Validate that value is equal to option."""
     return make_validator(
-        lambda value: value == option, lambda value: f"{value} does not match {option}."
+        lambda value: value == option,
+        lambda value, record_type, friendly_name, item_num: error_func(option)
+        if error_func
+        else f"{record_type}: {value} does not match {option}.",
     )
 
 
 def notMatches(option):
     """Validate that value is not equal to option."""
     return make_validator(
-        lambda value: value != option, lambda value: f"{value} matches {option}."
+        lambda value: value != option,
+        lambda value, record_type, friendly_name, item_num: f"{record_type}: {value} matches {option}."
     )
 
 
 def oneOf(options=[]):
     """Validate that value does not exist in the provided options array."""
     return make_validator(
-        lambda value: value in options, lambda value: f"{value} is not in {options}."
+        lambda value: value in options,
+        lambda value, record_type, friendly_name, item_num: f"{record_type}: {value} is not in {options}."
     )
 
 
 def notOneOf(options=[]):
     """Validate that value exists in the provided options array."""
     return make_validator(
-        lambda value: value not in options, lambda value: f"{value} is in {options}."
+        lambda value: value not in options,
+        lambda value, record_type, friendly_name, item_num: f"{record_type}: {value} is in {options}."
     )
 
 
 def between(min, max):
     """Validate value, when casted to int, is greater than min and less than max."""
     return make_validator(
         lambda value: int(value) > min and int(value) < max,
-        lambda value: f"{value} is not between {min} and {max}.",
+        lambda value, record_type, friendly_name, item_num: f"{record_type}: {value} is not between {min} and {max}.",
     )
 
 
-def hasLength(length, error_func=None):
+def recordHasLength(length):
     """Validate that value (string or array) has a length matching length param."""
     return make_validator(
         lambda value: len(value) == length,
-        lambda value: error_func(value, length)
-        if error_func
-        else f"Value length {len(value)} does not match {length}.",
+        lambda value, record_type,
```

Comment:
@elipe17 when the header record is missing, the feedback report raises the following 2  error messages, and the first one is misleading. I think this is because the system assumes the first record is the header record, but its actually T1 (test file below)

```shell=
HEADER record length is 117 characters but must be 23.
Your file does not begin with a HEADER record.
```

[2838.txt](https://github.com/raft-tech/TANF-app/files/14668185/2838.txt)

Reply:
@ADPennington that is the correct sequence of errors (see screenshot below). The preparsing validators do a length check first and then check if the record starts with HEADER. We can change the order of those errors if you'd like to see the second error occur first. 
![Screenshot 2024-03-21 at 9 07 10 AM](https://github.com/raft-tech/TANF-app/assets/125676261/4be0de37-5b69-4c6e-93e8-786b970fcb12)

Reply:
i think what's tricky about this is that we say the file is missing a header record, and then we also say your header record isn't the correct length, which could be confusing for an end user. for someone unfamiliar, would they understand that the length-related error message doesn't really need to be addressed if they add a header record to the file? @elipe17 @reitermb

Reply:
I think this could be confusing, but probably on the level of generating a support email vs acting incorrectly on the file. That said, would we be able to throw in a logic stop to prevent generating further errors once we serve up a [record] is missing error? I'm not sure if this can happen on anything other than header/trailer? @elipe17

Reply:
@reitermb, @ADPennington to add that logic would in my opinion be very un-generic because of the way we would have to track that. We would have to check for specific validators and their potential error messages which would breed a lot of one off code and some serious code smells. I think if this doesn't effect the user in a logic breaking way we should generate another ticket for this to be tracked and get this closed.

Reply:
@reitermb could we add guidance in #2847 on this? this is the first key check inside the file (i.e. files must have a header record) so feels worthy of being included in the TDP knowledge center. cc: @elipe17

Reply:
> @reitermb could we add guidance in #2847 on this? this is the first key check inside the file (i.e. files must have a header record) so feels worthy of being included in the TDP knowledge center. cc: @elipe17 

Yes, I think we can definitely include this alongside the broader explanation of what the header is. I think that's very relevant to the layout of the file generally.